### PR TITLE
cmake: Wip cmake compile flags

### PR DESCRIPTION
**Pull request by alimaredia**
Removes FORTIFY_SOURCE compile warnings and sets default CMAKE_BUILD_TYPE

### CODE REVIEW

**Comment by alimaredia** [comment]
@cbodley @liewegas @dillaman @adamemerson  care to chime in? Feel free to add anyone else to this conversation. Is the Automake default "Release"?

**Comment by dillaman** [comment]
```do_autogen.sh``` by defaults to debug enabled otherwise ```configure``` defaults to debug disabled -- both with optimizations (i.e. release) enabled.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -21,12 +21,16 @@ set(CMAKE_ASM_COMPILER  ${PROJECT_SOURCE_DIR}/src/yasm-wrapper)
...
 include(CheckCCompilerFlag)
-CHECK_C_COMPILER_FLAG("-Wp,-D_FORTIFY_SOURCE=2" HAS_FORTIFY_SOURCE)
+CHECK_C_COMPILER_FLAG("-Wp,-D_FORTIFY_SOURCE=2 -Werr" HAS_FORTIFY_SOURCE)
 if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
-if(HAS_FORTIFY_SOURCE)
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wp,-D_FORTIFY_SOURCE=2")
-endif()
+  if(HAS_FORTIFY_SOURCE)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wp,-D_FORTIFY_SOURCE=2")
+  endif()
```
Comment (cbodley) [inline src/CMakeLists.txt]: don't need the check for Debug anymore

and can you add the -U_FORTIFY_SOURCE flags from @dx9's pr?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -21,12 +21,14 @@ set(CMAKE_ASM_COMPILER  ${PROJECT_SOURCE_DIR}/src/yasm-wrapper)
...
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -rdynamic -Wall -Wtype-limits -Wignored-qualifiers -Winit-self -Wpointer-arith -Werror=format-security -fno-strict-aliasing -fsigned-char -fPIC")
 
+if(NOT CMAKE_BUILD_TYPE) 
+  set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Default BUILD_TYPE is RelWithDebInfo, other options are: Debug Release MinSizeRel.") 
+endif(NOT CMAKE_BUILD_TYPE)
+
 include(CheckCCompilerFlag)
-CHECK_C_COMPILER_FLAG("-Wp,-D_FORTIFY_SOURCE=2" HAS_FORTIFY_SOURCE)
-if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
+CHECK_C_COMPILER_FLAG("-Wp,-U_FORTIFY_SOURCE, -D_FORTIFY_SOURCE=2 -Werr" HAS_FORTIFY_SOURCE)
```
Comment (alimaredia) [inline src/CMakeLists.txt]: @dx9 is this the way you would have had from  your deleted PR?

**Comment by mattbenjamin** [comment]
@alimaredia remove fortify warnings?  at last!!

**Comment by alimaredia** [comment]
@mattbenjamin you read it right!

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -21,12 +21,14 @@ set(CMAKE_ASM_COMPILER  ${PROJECT_SOURCE_DIR}/src/yasm-wrapper)
...
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -rdynamic -Wall -Wtype-limits -Wignored-qualifiers -Winit-self -Wpointer-arith -Werror=format-security -fno-strict-aliasing -fsigned-char -fPIC")
 
+if(NOT CMAKE_BUILD_TYPE) 
+  set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Default BUILD_TYPE is RelWithDebInfo, other options are: Debug Release MinSizeRel.") 
+endif(NOT CMAKE_BUILD_TYPE)
+
 include(CheckCCompilerFlag)
-CHECK_C_COMPILER_FLAG("-Wp,-D_FORTIFY_SOURCE=2" HAS_FORTIFY_SOURCE)
-if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
+CHECK_C_COMPILER_FLAG("-Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=2 -Werr" HAS_FORTIFY_SOURCE)
```
Comment (dx9) [inline src/CMakeLists.txt]: @alimaredia The compiler check fails:

/usr/bin/cc    -DHAS_FORTIFY_SOURCE   -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=2 -Werr -o CMakeFiles/cmTC_b97c1.dir/src.c.o   -c /home/build/ceph/test-cmake/build/CMakeFiles/CMakeTmp/src.c
cc: error: unrecognized command line option '-Werr'
make[1]: *** [CMakeFiles/cmTC_b97c1.dir/build.make:66: CMakeFiles/cmTC_b97c1.dir/src.c.o] Error 1
make[1]: Leaving directory '/home/build/ceph/test-cmake/build/CMakeFiles/CMakeTmp'
make: *** [Makefile:126: cmTC_b97c1/fast] Error 2

Changing -Werr to -Werror fixes it. 

Thanks!

Comment (tchaikov) [inline src/CMakeLists.txt]: @alimaredia can we use `-Werror` instead?

**Comment by tchaikov** [comment]
other than `s/-Werr/-Werror/`, lgtm.

**Comment by alimaredia** [comment]
@tchaikov @cbodley "-werror" change is in

File: src/CMakeLists.txt (hunk 1 above)
Comment (tchaikov) [inline src/CMakeLists.txt]: > don't need the check for Debug anymore

@cbodley we still need it? as `CHECK_C_COMPILER_FLAG()` does not respect `CMAKE_BUILD_TYPE`, i think. and fortify cflags does not work with debug build.

**Pull request merged**